core/src/main/java/org/elasticsearch/action/support/replication/ReplicatedOperation.java

```diff
@@ -0,0 +1,334 @@
+        finalResponse = primaryResponse.v1();
+        ReplicaRequest replicaRequest = primaryResponse.v2();
+        replicaRequest.primaryTerm(primary.term());
+        if (logger.isTraceEnabled()) {
+            logger.trace("[{}] op [{}] completed on primary for request [{}]", primary.routingEntry().shardId(), opType, request);
+        }
+        // we have to get a new state after successfully indexing into the primary in order to honour recovery semantics.
+        // we have to make sure that every operation indexed into the primary after recovery start will also be replicated
+        // to the recovery target. If we use an old cluster state, we may miss a relocation that has started since then.
+        // If the index gets deleted after primary operation, we skip replication
+        List<ShardRouting> shards = getShards(request.shardId(), clusterStateSupplier.get());
+        final String localNodeId = primary.routingEntry().currentNodeId();
+        for (final ShardRouting shard : shards) {
+            if (executeOnReplicas == false || shard.unassigned()) {
+                if (shard.primary() == false) {
+                    totalShards.incrementAndGet();
+                }
+                continue;
+            }
+
+            if (shard.currentNodeId().equals(localNodeId) == false) {
+                performOnReplica(shard, replicaRequest);
+            }
+
+            if (shard.relocating() && shard.relocatingNodeId().equals(localNodeId) == false) {
+                performOnReplica(shard.buildTargetRelocatingShard(), replicaRequest);
+            }
+        }
+
+        // mark primary as successful, decrement pending and finish (if there are no replicas, or those are done)
+        totalShards.incrementAndGet();
+        successfulShards.incrementAndGet();
+        decPendingAndFinishIfNeeded(); // incremented in the beginning of this method
+    }
+
+    private void performOnReplica(final ShardRouting shard, final ReplicaRequest replicaRequest) {
+        if (logger.isTraceEnabled()) {
+            logger.trace("[{}] sending op [{}] to replica {} for request [{}]", shard.shardId(), opType, shard, replicaRequest);
+        }
+
+        totalShards.incrementAndGet();
+        pendingShards.incrementAndGet();
+        replicasProxy.performOn(shard, replicaRequest, new ActionListener<TransportResponse.Empty>() {
+            @Override
+            public void onResponse(TransportResponse.Empty empty) {
+                successfulShards.incrementAndGet();
+                decPendingAndFinishIfNeeded();
+            }
+
+            @Override
+            public void onFailure(Throwable replicaException) {
+                logger.trace("[{}] failure while performing [{}] on replica {}, request [{}]", replicaException, shard.shardId(), opType,
+                        shard, replicaRequest);
+                if (ignoreReplicaException(replicaException)) {
```

Comment:
why are you looking here at the specific exception returned by transport but adding a special parameter `onPrimaryDemoted ` to `replicasProxy.failShard(...)`?